File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] nger = nextGreaterElementRight(nums2);
        
        int[] res = new int[nums1.length];
        
        for(int i = 0; i<nums2.length; i++){
            int ele = nums2[i];
            for(int j=0; j<nums1.length; j++){
                if(nums1[j] == ele){
                    res[j] = nger[i];
                    break;
                }
            }
        }
        
        return res;
    }
    
    public int[] nextGreaterElementRight(int[] arr){
        int len = arr.length;
        Stack<Integer> st = new Stack<>();
        int []nger = new int[len];
        
        for(int i = 0; i<len; i++){
            int ele = arr[i];
            while(st.size() > 0 && ele > arr[st.peek()]){
                int idx = st.pop();
                nger[idx] = ele;
            }
            
            st.push(i);
        }
        while(st.size() > 0){
            nger[st.pop()] = -1;
        }
        
        return nger;
    }
}

File: 0496-next-greater-element-i/0496-next-greater-element-i.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, nums2) {
    const map = {};
    const st = []; // Stack => LIFO,
    
    for(let i=0; i<nums2.length; i++){
        let ele = nums2[i];
        // Whenever, the ele > nums2[st.peek()]
        while(st.length > 0 && ele > nums2[st[st.length -1]]){
            map[nums2[st.pop()]] = ele;
        }
        // Pushing the index into the Stack:
        st.push(i);
    }
    // If Stack Still Has elements:
    while(st.length > 0){
        map[nums2[st.pop()]] = -1;
    }
    // console.log(map);
    const res = [];
    for(let i=0; i<nums1.length; i++){
        let ele = nums1[i];
        let nger = map[ele];
        res[i] = nger;
    }
    return res;
};
